[docs] intro for query tuning (#8796)
yugabyte/yugabyte-db: docs/content/latest/explore/query-1-performance/query-tuning-intro.md

```

```output
Master UUID                        RPC Host/Port           State       Role
a637b88dfc0c4476862ca79872d763d7   172.158.22.229:7100     ALIVE       LEADER
968eb4a7a2e24c3ab95929d3a888ed05   172.158.39.23:7100      ALIVE       FOLLOWER
44e13e096955412a88c6ffd65a1e241c   172.158.50.212:7100     ALIVE       FOLLOWER
```

```sh
$ ./bin/yb-admin -init_master_addrs=$(hostname):7100 list_all_tablet_servers
```

```output
Tablet Server UUID               RPC Host/Port       Heartbeat delay Status   Reads/s  
d80150e0eeda4477a231968440dd89a9 172.158.50.212:9100 0.20s           ALIVE    .. 
721661fd5d2044179e03707a862aa578 172.158.22.229:9100 0.39s           ALIVE    .. 
54e17c914c0f4679aa7b07a9c7a8ddf5 172.158.39.23:9100  0.48s           ALIVE    ..
```

Next, make sure enough disk space is available, elementary components such as CPU, disk, and the network do not report errors, and the operating system does not report any malfunction.

For more information on troubleshooting cluster- and node-level issues, refer to [Troubleshoot YugabyteDB](../../../troubleshoot).

Once you know the entire cluster is running correctly, you can move to statement tuning.

> ## Tuning in YSQL / Postgres

Reviewer: Don't use slashes, especially in titles. Also, spell out PostgreSQL every time, as that's the real name.

Except for here, where you might want to drop it entirely and just say "Tuning in YSQL".

Author: fixed

> Each node running the YB-TServer process also runs an instance of Postgres (running as the YugabyteDB YSQL API). (This is often referred to as 'stateless Postgres', because each instance of Postgres runs independently.)

Reviewer: This feels off from the description in /latest/explore/ysql-language-features/ — maybe try to get the two to line up a bit more.

Author: fixed

> To investigate a specific apparently slow running query, you first need to determine which node the query is running on. 

Reviewer: "To identify the node on which the query is running, you can..."

Author: Waiting to hear back from @fritshoogland-yugabyte  or @epratt-yb on how to identify problem nodes in Core

{{< note title="Note" >}}

> Remember that even for a single query running on one YSQL node, the whole cluster can show activity, because the storage layer is distributed over all nodes.

Reviewer: Is a "YSQL node" actually Yugabyte terminology? It does not appear anywhere in the docs but here. I wonder if you want to say "node" for all of these.

Author: Fixed

{{< /note >}}

Once you identify the correct YSQL instance, YugabyteDB YSQL provides two views you can use to identify SQL statements and their performance characteristics:

> * `pg_stat_activity` - provides an overview, including SQL, of current activity. Apart from the query text, it does not provide information about the SQL executed; rather it shows the current state of the Postgres backend.
> 
> * `pg_stat_statements` - provides historic information about executed SQL statements. Currently, no 'block'/'blk' information is recorded (which is logical and physical IO information), because the Postgres IO codepath is not used.

Reviewer: You don't need the dashes here. "I/O" is always spelled with a slash, though.
```suggestion
* `pg_stat_activity` provides an overview, including SQL, of current activity. Apart from the query text, it does not provide information about the SQL executed; rather it shows the current state of the Postgres backend.

* `pg_stat_statements` provides historic information about executed SQL statements. Currently, no 'block'/'blk' information is recorded (which is logical and physical I/O information), because the Postgres I/O codepath is not used.
```

Author: fixed

Once you identify a problematic SQL statement, the next step is to look at the execution plan. An execution plan is the order and method of execution to perform the work requested in the SQL statement. This plan is generated by the database planner.

YugabyteDB uses PostgreSQL’s cost-based optimizer, which estimates the costs of each possible execution plan for an SQL statement. The planner calculates the cheapest plan to execute a query, assuming that is also the fastest and the best plan, based on statistics derived from the database table and supporting structures like secondary indexes. The execution plan with the lowest cost finally is executed. 

{{< note title="Note" >}}

Currently, YugabyteDB does not perform size-specific costing, and assumes every plan generates a thousand rows, which translates to a cost of a thousand. 

{{< /note >}}

To view the execution plan, use the `EXPLAIN` statement. You can use `EXPLAIN` in two modes:

* The default `EXPLAIN` mode shows the plan that the planner created and the predicted cost.

* `EXPLAIN ANALYZE` mode executes the query, measures the time and actual work done, and adds these statistics to the execution plan as 'actual'.

> Both the predicted and actual statistics have their uses: 
> 
> * The predicted statistics are what the planner uses, and assuming the statistics don't change, the planner will come up with the same execution plan the next time the query is run, even if it caused performance issues in the past.
> 
> * The actual statistics show how much and what work was actually performed, and thus are useful to see the effectiveness of the execution plan. 

Reviewer: So many bullets. Let's take some out.
```suggestion
Both the predicted and actual statistics have their uses. The _predicted_ statistics are what the planner uses, and assuming the statistics don't change, the planner will come up with the same execution plan the next time the query is run, even if it caused performance issues in the past. The _actual_ statistics show how much and what work was actually performed, and thus are useful to see the effectiveness of the execution plan. 
```

Author: fixed

{{< tip title="Tip" >}}

Complex execution plans can be difficult to read or understand. Use online tools to calculate and visualize execution plan timelines. For example, <https://explain.dalibo.com>.

{{< /tip >}}

## Performance tuning tools in YugabyteDB

> 

Reviewer: In general, try to avoid having two headings in a row, unless the introductory text you would put in between is really useless and contrived.

> ### Find slow queries using pg_stat_statements

Reviewer: To make the right nav nicer, consider shortening this one and the next.
```suggestion
### Find slow queries
```

Author: fixed

Use the `pg_stat_statements` extension to get statistics on past queries. Using `pg_stat_statements`, you can investigate queries by userid and dbid, calls, rows, and min, max, mean, standard deviation and total time.

> The [`pg_stat_statements`](https://www.postgresql.org/docs/11/pgstatstatements.html) extension module is installed by default, but must be enabled for a database before you can query the `pg_stat_statements` view.

Reviewer: You're linking to the PostgreSQL site before you're linking to our own docs. That's probably backwards :)

Author: fixed

```sql
CREATE EXTENSION if not exists pg_stat_statements;
```

To get the output of `pg_stat_statements` in JSON format, visit `https://<yb-tserver-ip>:13000/statements` in your web browser, where `<yb-tserver-ip>` is the IP address of any YB-TServer node in your cluster.

For more information, refer to [Get query statistics using `pg_stat_statements`](../pg-stat-statements).

> ### View live queries using pg_stat_activity

Reviewer: (Same as previous heading suggestion)
```suggestion
### View live queries
```

Author: fixed

Use the `pg_stat_activity` view to get information on currently running tasks. Using `pg_stat_activity` you can identify inactive, active, and long time active sessions, and get process information and the current query.

To get the output of `pg_stat_activity` in JSON format, visit `https://<yb-tserver-ip>:13000/rpcz` in your web browser, where `<yb-tserver-ip>` is the IP address of any YB-TServer node in your cluster.

For more information, refer to [Viewing live queries with `pg_stat_activity`](../pg-stat-activity).

> ### Understand what your queries are doing with EXPLAIN

Reviewer: There's a bunch of EXPLAIN earlier on the page. Can you consolidate?

Author: I could - I was leaving a section on each tool, even though the first 3 are discussed in the earlier overview section, to give each one equal prominence in the page TOC, but they could be removed if you think it's confusing

> ### Understand what your queries are doing with EXPLAIN

Reviewer: Also, maybe shorten this one for the right nav too? Something like this?
```suggestion
### Use EXPLAIN for queries
```

Author: fixed

Like PostgreSQL, YugabyteDB provides the `EXPLAIN` statement to show the query execution plan generated by YSQL for a given SQL statement. Using `EXPLAIN`, you can discover where in the query plan the query is spending most of its time, and using this information, decide on the best approach for improving query performance. This could include strategies such as adding an index or changing index sort order.

For more information, refer to [Analyzing queries with EXPLAIN](../explain-analyze).

## Advanced tools

> ### Turn on slow query logging

Reviewer: ```suggestion
### Log slow queries
```

Author: fixed

```sh
$ ./bin/yb-tserver --ysql_log_min_duration_statement 1000
```

Example log output:

```output
2021-05-26 21:13:49.701 EDT [73005] LOG:  duration: 34.378 ms  statement: SELECT c.oid,
        n.nspname,
        c.relname
    FROM pg_catalog.pg_class c
        LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
    WHERE c.relname OPERATOR(pg_catalog.~) '^(products)$'
        AND pg_catalog.pg_table_is_visible(c.oid)
    ORDER BY 2, 3;
```

Results are written to the current `postgres*log` file. For information on the YB-TServer logs, refer to [YB-TServer logs](../../../troubleshoot/nodes/check-logs/#yb-tserver-logs).

{{< note title="Note" >}}

Depending on the database and the work being performed, long-running queries don't necessarily need to be optimized.

Ensure that the threshold is high enough so that you don't flood the `postgres*log` log files.

{{< /note >}}

For more information on flags for configuring the YB-TServer server, refer to [YSQL Flags](../../../reference/configuration/yb-tserver/#ysql-flags).

> ### Optimize YSQL queries using pg_hint_plan

Reviewer: ```suggestion
### Use a hint plan
```

Author: fixed
